accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,918 @@
+var jValue = new JsonValue<JsonElement>(jsonElement);
+var jObject = new JsonObject(jsonElement);
+var jArray = new JsonArray(jsonElement);
+```
+
+# Programming model notes
+## Why `JsonValue` and not `JsonNumber` + `JsonString` + `JsonBoolean`?
+The JSON primitives (number, string, true, false) do **not** have their own `JsonNode`-derived type like `JsonNumber`, `JsonString` and `JsonBoolean`. Instead, a common `JsonValue` class represents them. This allows a JSON number, for example, to serialize and deserialize as either a CLR `String` or a `Double` depending on usage and options.
+
+## Cast operators for JsonValue<T>
+Although explicit or implicit cast operators are not required for functionality, based on usability feedback requesting a terse syntax, the use of both explicit cast operators (from a `JsonValue<T>` to known primitives) and implicit cast operators (from known primitives to a `JsonValue<T>`) are supported.
+
+Note that cast operators do not work in all languages such as F#.
+
+Explicit (no operators):
+```cs
+JsonArray jArray = ...
+jArray[0] = new JsonValue<string>("hello"); 
+jArray[1] = new JsonElement<MyCustomDataType>(myCustomDataType);
+
+string s = jArray[0].GetValue<string>();
+MyCustomDataType m = jArray[1].GetValue<MyCustomDataType>();```
+```
+with operators:
+```cs
+JsonArray jArray = ...
+jArray[0] = "hello"; // Uses implicit operator.
+jArray[1] = new JsonElement<MyCustomDataType>(myCustomDataType); // No implicit operator for custom types (unless it adds its own)
+
+string s = (string)jArray[0]; // Uses explicit operator.
+MyCustomDataType m = jArray[1].GetValue<MyCustomDataType>(); // no explicit operator for custom types  (unless it adds its own).
+```
+with dynamic:
+```cs
+JsonArray jArray = ...
+jArray[0] = "hello"; // Uses implicit operator.
+jArray[1] = myCustomDataType; // Possible through dynamic.
+
+string s = jArray[0]; // Possible through dynamic.
+MyCustomDataType m = jArray[1]; // Possible through dynamic.
+```
+
+## Specifying the `JsonSerializerOptions` instance
+The `JsonSerializerOptions` instance needs to be specified in order for any run-time added custom converters to be found, to support `PropertyNameCaseInsensitive` for property lookup, and other options such as `NumberHandling`.
+
+It is cumbersome and error-prone to specify the options instance for every new element or property. Consider the verbose syntax:
+```cs
+// Verbose syntax (not recommended, but still possible in the API)
+var jObject = new JsonObject(options)
+{
+    ["MyString"] = new JsonValue<string>("Hello!", options),
+    ["MyBoolean"] = new JsonValue<bool>(false, options),
+    ["MyArray"] = new JsonArray(options)
+    {
+        new JsonValue(2, options),
+        new JsonValue(3, options),
+        new JsonValue(42, options)
+    },
+}
+```
+and the terse syntax which omits the options instance for non-root members. When a property or element is added, the options instance is set to the parent's instance:
```

Comment:
Wouldn't that lead to confusion if the same node instance is assigned to trees with different configs? It doesn't seem too unlikely to me (e.g. when some application caches a `JsonValue` in a static).

Reply:
Per offline discussion, to support caching or related scenarios where multiple options are used, I added additional clarity in that section that allows different option instances to be specified within a tree, but during a given serialization, only the options from the node passed to serialize()\ToJsonString()\etc is used.